# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/loader/WorkerLoadContext.h
# Commit: 5f483f0b842a
# Full Hash: 5f483f0b842afdfad82ea041256f9d8bce37f421
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-11-02 04:41:32
# Regressor Bug: 1797327
# File Overlap Count: 3
# Description:
#   Bug 1797327 - Use RefPtrs for all WorkerLoadContexts; r=asuth
#   
#   As WorkerLoadContexts now inherit from a non-CC'd loadContextBase, we have two outcomes.
#   1) we need to break cycles with ScriptLoadRequests manually, so that ScriptLoadRequests can be collected (ScriptLoadRequests must be CC'd).
#   2) we can now have refptrs to WorkerLoadContexts in the CacheLoadHandler and NetworkLoadHandler classes, and remove any remaining raw pointers to ScriptLoadRequest/WorkerLoadContext. There are cases where the NetworkLoadHandler or CacheLoadHandler might outlive the Worker Loader, so having refpointers here should help us recover in those cases.
# ==============================================================================

diff -r 7ade701781f5 -r 5f483f0b842a dom/workers/loader/WorkerLoadContext.h
--- a/dom/workers/loader/WorkerLoadContext.h	Tue Nov 01 18:03:34 2022 +0000
+++ b/dom/workers/loader/WorkerLoadContext.h	Tue Nov 01 18:03:34 2022 +0000
@@ -80,10 +80,13 @@
 
   explicit WorkerLoadContext(Kind aKind, const Maybe<ClientInfo>& aClientInfo);
 
+  void SetRequest(JS::loader::ScriptLoadRequest* aRequest) override {
+    LoadContextBase::SetRequest(aRequest);
+    mIsTopLevel = aRequest->IsTopLevel() && (mKind == Kind::MainScript);
+  }
+
   // Used to detect if the `is top-level` bit is set on a given module.
-  bool IsTopLevel() {
-    return mRequest->IsTopLevel() && (mKind == Kind::MainScript);
-  };
+  bool IsTopLevel() { return mIsTopLevel; };
 
   static Kind GetKind(bool isMainScript, bool isDebuggerScript) {
     if (isDebuggerScript) {
@@ -99,6 +102,7 @@
   Maybe<bool> mMutedErrorFlag;
   nsresult mLoadResult = NS_ERROR_NOT_INITIALIZED;
   bool mLoadingFinished = false;
+  bool mIsTopLevel = true;
   Kind mKind;
   Maybe<ClientInfo> mClientInfo;
 